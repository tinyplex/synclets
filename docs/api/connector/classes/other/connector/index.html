<!DOCTYPE html><html lang="en"><head><meta name="viewport" content="width=device-width,initial-scale=1"><link rel="preload" as="image" href="/favicon.svg"><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-14BZ56H85M"></script><link rel="dns-prefetch" href="https://www.googletagmanager.com/"><link href="https://www.googletagmanager.com/gtag/js?id=G-14BZ56H85M" rel="preload" as="script"><link rel="preload" as="font" href="/fonts/inter.woff2" type="font/woff2" crossorigin="anonymous"><link rel="preload" as="font" href="/fonts/inconsolata.woff2" type="font/woff2" crossorigin="anonymous"><link rel="preload" as="font" href="/fonts/shantell.woff2" type="font/woff2" crossorigin="anonymous"><title>Connector | Synclets</title><meta name="description" content=""><meta property="og:type" content="website"><meta property="og:title" content="Connector | Synclets"><meta property="og:description" content=""><meta property="og:url" content="https://synclets.org/api/connector/classes/other/connector/"><meta property="og:image" content="https://synclets.org/favicon_pad.png"><meta name="twitter:card" content="summary"><meta name="twitter:title" content="Connector | Synclets"><meta name="twitter:description" content=""><meta name="twitter:site" content="@tinybasejs"><meta name="twitter:image" content="https://synclets.org/favicon_pad.png"><link rel="icon" type="image/svg+xml" href="/favicon.svg"><link rel="canonical" href="https://synclets.org/api/connector/classes/other/connector/"><link rel="stylesheet" href="/css/index.css"><script src="/js/app.js"></script></head><body><header><a href="/"><img src="/favicon.svg" alt="Synclets logo"><span>Synclets</span></a><nav><ul><li><a href="/guides/">Guides</a></li><li><a href="/api/">API</a></li><li><a href="https://github.com/tinyplex/synclets">GitHub</a></li></ul></nav><span id="dark"></span></header><main><nav><ul><li id="root" class="parent open"><span></span><a href="/">Synclets</a><ul><li id="G" class="parent"><span></span><a href="/guides/">Guides</a></li><li id="api" class="parent open"><span></span><a href="/api/">API</a><ul><li id="e"><span></span><a href="/api/the-essentials/">The Essentials</a></li><li id="s" class="parent"><span></span><a href="/api/synclets/"><code>synclets</code></a></li><li id="c" class="parent open"><span></span><a href="/api/connector/"><code>connector</code></a><ul><li id="c/C" class="parent open"><span></span><a href="/api/connector/classes/">Classes</a><ul><li id="C2" class="current parent open"><span></span><a href="/api/connector/classes/other/connector/"><code>Connector</code></a><ul><li id="C2/C3" class="parent"><span></span><a href="/api/connector/classes/other/connector/constructors/">Constructors</a></li><li id="C2/M2" class="parent"><span></span><a href="/api/connector/classes/other/connector/methods/">Methods</a></li></ul></li></ul></li><li id="c/TA2" class="parent"><span></span><a href="/api/connector/type-aliases/">Type Aliases</a></li></ul></li><li id="c4" class="parent"><span></span><a href="/api/connector-value/"><code>connector-value</code></a></li><li id="t" class="parent"><span></span><a href="/api/transport/"><code>transport</code></a></li><li id="t2" class="parent"><span></span><a href="/api/transport-memory/"><code>transport-memory</code></a></li></ul></li></ul></li></ul></nav><article><nav><ul><li><a href="/">Synclets</a></li><li><a href="/api/">API</a></li><li><a href="/api/connector/"><code>connector</code></a></li><li><a href="/api/connector/classes/">Classes</a></li><li><a href="/api/connector/classes/other/connector/"><code>Connector</code></a></li></ul></nav><section class="s1" id="/api/connector/classes/other/connector/" data-id="C2"><h1><code>Connector</code></h1><section class="s2" id="/api/connector/classes/other/connector/constructors/" data-id="C2/C3"><h2>Constructors</h2><ul><li><a href="/api/connector/classes/other/connector/constructors/other/constructor/"><code>constructor</code></a></li></ul></section><section class="s2" id="/api/connector/classes/other/connector/methods/" data-id="C2/M2"><h2>Methods</h2><ul><li><a href="/api/connector/classes/other/connector/methods/other/connect/"><code>connect</code></a></li><li><a href="/api/connector/classes/other/connector/methods/other/disconnect/"><code>disconnect</code></a></li></ul></section></section></article><aside></aside></main><footer><nav><a id="gh" href="https://github.com/tinyplex/synclets" target="_blank" rel="noreferrer">GitHub</a><a id="bs" href="https://bsky.app/profile/tinybase.bsky.social" target="_blank" rel="noreferrer">Bluesky</a><a id="tw" href="https://x.com/tinybasejs" target="_blank" rel="noreferrer">X / Twitter</a><a id="dc" href="https://discord.com/invite/mGz3mevwP8" target="_blank" rel="noreferrer">Discord</a></nav><nav><a href="/">Synclets <span id="version"></span></a> Â© 2025-</nav></footer><script>window.dataLayer=window.dataLayer||[];function g(){dataLayer.push(arguments);}g('js',new Date());g('config','G-14BZ56H85M');</script></body></html>