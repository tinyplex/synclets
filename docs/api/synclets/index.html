<!DOCTYPE html><html lang="en"><head><meta name="viewport" content="width=device-width,initial-scale=1"><link rel="preload" as="image" href="/favicon.svg"><script async="" src="https://www.googletagmanager.com/gtag/js?id=G-14BZ56H85M"></script><link rel="dns-prefetch" href="https://www.googletagmanager.com/"><link href="https://www.googletagmanager.com/gtag/js?id=G-14BZ56H85M" rel="preload" as="script"><link rel="preload" as="font" href="/fonts/inter.woff2" type="font/woff2" crossorigin="anonymous"><link rel="preload" as="font" href="/fonts/inconsolata.woff2" type="font/woff2" crossorigin="anonymous"><link rel="preload" as="font" href="/fonts/shantell.woff2" type="font/woff2" crossorigin="anonymous"><title>synclets | Synclets</title><meta name="description" content="The synclets module."><meta property="og:type" content="website"><meta property="og:title" content="synclets | Synclets"><meta property="og:description" content="The synclets module."><meta property="og:url" content="https://synclets.org/api/synclets/"><meta property="og:image" content="https://synclets.org/favicon_pad.png"><meta name="twitter:card" content="summary"><meta name="twitter:title" content="synclets | Synclets"><meta name="twitter:description" content="The synclets module."><meta name="twitter:site" content="@tinybasejs"><meta name="twitter:image" content="https://synclets.org/favicon_pad.png"><link rel="icon" type="image/svg+xml" href="/favicon.svg"><link rel="canonical" href="https://synclets.org/api/synclets/"><link rel="stylesheet" href="/css/index.css"><script src="/js/app.js"></script></head><body><header><a href="/"><img src="/favicon.svg" alt="Synclets logo"><span>Synclets</span></a><nav><ul><li><a href="/guides/">Guides</a></li><li><a href="/api/">API</a></li><li><a href="https://github.com/tinyplex/synclets">GitHub</a></li></ul></nav><span id="dark"></span></header><main><nav><ul><li id="root" class="parent open"><span></span><a href="/">Synclets</a><ul><li id="G" class="parent"><span></span><a href="/guides/">Guides</a></li><li id="api" class="parent open"><span></span><a href="/api/">API</a><ul><li id="e"><span></span><a href="/api/the-essentials/">The Essentials</a></li><li id="s" class="current parent open"><span></span><a href="/api/synclets/"><code>synclets</code></a><ul><li id="s/I" class="parent"><span></span><a href="/api/synclets/interfaces/">Interfaces</a></li><li id="s/F" class="parent"><span></span><a href="/api/synclets/functions/">Functions</a></li><li id="s/TA" class="parent"><span></span><a href="/api/synclets/type-aliases/">Type Aliases</a></li></ul></li><li id="c2" class="parent"><span></span><a href="/api/connector-base/"><code>connector/base</code></a></li><li id="c7" class="parent"><span></span><a href="/api/connector-file/"><code>connector/file</code></a></li><li id="t" class="parent"><span></span><a href="/api/transport-memory/"><code>transport/memory</code></a></li><li id="u" class="parent"><span></span><a href="/api/utils/"><code>utils</code></a></li></ul></li></ul></li></ul></nav><article><nav><ul><li><a href="/">Synclets</a></li><li><a href="/api/">API</a></li><li><a href="/api/synclets/"><code>synclets</code></a></li></ul></nav><section class="s1" id="/api/synclets/" data-id="s"><h1><code>synclets</code></h1><p>The <code>synclets</code> module.</p><section class="s2"><h2>Since</h2><p>v0.0.0</p></section><section class="s2" id="/api/synclets/interfaces/" data-id="s/I"><h2>Interfaces</h2><ul><li><a href="/api/synclets/interfaces/other/connector/"><code>Connector</code></a></li><li><a href="/api/synclets/interfaces/other/synclet/"><code>Synclet</code></a></li><li><a href="/api/synclets/interfaces/other/transport/methods/other/log/"><code>log</code></a> method</li></ul></section><section class="s2" id="/api/synclets/functions/" data-id="s/F"><h2>Functions</h2><ul><li><a href="/api/synclets/functions/other/getpacketfromparts/"><code>getPacketFromParts</code></a></li><li><a href="/api/synclets/functions/other/getpartsfrompacket/"><code>getPartsFromPacket</code></a></li><li><a href="/api/synclets/functions/other/createconnector/"><code>createConnector</code></a></li><li><a href="/api/synclets/functions/other/createsynclet/"><code>createSynclet</code></a></li><li><a href="/api/synclets/functions/other/createtransport/"><code>createTransport</code></a></li></ul></section><section class="s2" id="/api/synclets/type-aliases/" data-id="s/TA"><h2>Type Aliases</h2><ul><li><a href="/api/synclets/type-aliases/other/address/"><code>Address</code></a></li><li><a href="/api/synclets/type-aliases/other/atom/"><code>Atom</code></a></li><li><a href="/api/synclets/type-aliases/other/connectorimplementations/"><code>ConnectorImplementations</code></a></li><li><a href="/api/synclets/type-aliases/other/connectoroptions/"><code>ConnectorOptions</code></a></li><li><a href="/api/synclets/type-aliases/other/context/"><code>Context</code></a></li><li><a href="/api/synclets/type-aliases/other/hash/"><code>Hash</code></a></li><li><a href="/api/synclets/type-aliases/other/loglevel/"><code>LogLevel</code></a></li><li><a href="/api/synclets/type-aliases/other/logger/"><code>Logger</code></a></li><li><a href="/api/synclets/type-aliases/other/node/"><code>Node</code></a></li><li><a href="/api/synclets/type-aliases/other/subnodes/"><code>SubNodes</code></a></li><li><a href="/api/synclets/type-aliases/other/syncletimplementations/"><code>SyncletImplementations</code></a></li><li><a href="/api/synclets/type-aliases/other/syncletoptions/"><code>SyncletOptions</code></a></li><li><a href="/api/synclets/type-aliases/other/timestamp/"><code>Timestamp</code></a></li><li><a href="/api/synclets/type-aliases/other/timestampandatom/"><code>TimestampAndAtom</code></a></li><li><a href="/api/synclets/type-aliases/other/tomb/"><code>Tomb</code></a></li><li><a href="/api/synclets/type-aliases/other/transportimplementations/"><code>TransportImplementations</code></a></li><li><a href="/api/synclets/type-aliases/other/transportoptions/"><code>TransportOptions</code></a></li></ul></section></section></article><aside></aside></main><footer><nav><a id="gh" href="https://github.com/tinyplex/synclets" target="_blank" rel="noreferrer">GitHub</a><a id="bs" href="https://bsky.app/profile/tinybase.bsky.social" target="_blank" rel="noreferrer">Bluesky</a><a id="tw" href="https://x.com/tinybasejs" target="_blank" rel="noreferrer">X / Twitter</a><a id="dc" href="https://discord.com/invite/mGz3mevwP8" target="_blank" rel="noreferrer">Discord</a></nav><nav><a href="/">Synclets <span id="version"></span></a> Â© 2025-</nav></footer><script>window.dataLayer=window.dataLayer||[];function g(){dataLayer.push(arguments);}g('js',new Date());g('config','G-14BZ56H85M');</script></body></html>