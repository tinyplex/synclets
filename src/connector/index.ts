import type {Connector as ConnectorDecl} from '../@types/connector/index.d.ts';

export class Connector implements ConnectorDecl {
  async connect(): Promise<void> {}

  async disconnect(): Promise<void> {}
}
