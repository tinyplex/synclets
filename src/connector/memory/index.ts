export {createMemoryDataConnector} from './data.ts';
export {createMemoryMetaConnector} from './meta.ts';
