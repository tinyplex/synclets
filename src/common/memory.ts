import {createDataConnector, createMetaConnector} from '@synclets';
import type {
  Address,
  Atom,
  Atoms,
  Context,
  Data,
  DataConnector,
  Meta,
  MetaConnector,
  Timestamp,
  Timestamps,
} from '@synclets/@types';
import {jsonParse, jsonString} from '@synclets/utils';
import {objDeepAction, objKeys} from './object.ts';
import {isEmpty} from './other.ts';

export const createMemoryDataConnector = async <Depth extends number>(
  depth: Depth,
  onChange?: (data: Data) => Promise<void>,
  initial?: Data,
): Promise<DataConnector<Depth>> => {
  const data: Data = initial ?? {};

  const readAtom = async (address: Address, _context: Context) =>
    objDeepAction(data, address, (parent, id) => parent[id]) as
      | Atom
      | undefined;

  const writeAtom = async (address: Address, atom: Atom, _context: Context) =>
    objDeepAction(
      data,
      address,
      (parent, id) => {
        parent[id] = atom;
        onChange?.(data);
      },
      true,
    );

  const removeAtom = async (address: Address, _context: Context) =>
    objDeepAction(
      data,
      address,
      (parent, id) => {
        delete parent[id];
        onChange?.(data);
      },
      true,
      true,
    );

  const readChildIds = async (address: Address, _context: Context) =>
    isEmpty(address)
      ? objKeys(data)
      : (objDeepAction(data, address, (parent, id) =>
          objKeys(parent[id] as Data),
        ) ?? []);

  const readAtoms = async (address: Address, _context: Context) =>
    objDeepAction(data, address, (parent, id) => parent[id] ?? {}) as Atoms;

  const connector = await createDataConnector(
    depth,
    {readAtom, writeAtom, removeAtom, readChildIds, readAtoms},
    {getData: async () => jsonParse(jsonString(data))},
  );

  return connector;
};

export const createMemoryMetaConnector = async <Depth extends number>(
  depth: Depth,
  onChange?: (meta: Meta) => Promise<void>,
  initial?: Meta,
): Promise<MetaConnector<Depth>> => {
  const meta: Meta = initial ?? {};

  const readTimestamp = async (address: Address, _context: Context) =>
    objDeepAction(meta, address, (parent, id) => parent[id]) as
      | Timestamp
      | undefined;

  const writeTimestamp = async (
    address: Address,
    timestamp: Timestamp,
    _context: Context,
  ) =>
    objDeepAction(
      meta,
      address,
      (parent, id) => {
        parent[id] = timestamp;
        onChange?.(meta);
      },
      true,
      true,
    );

  const readChildIds = async (address: Address, _context: Context) =>
    isEmpty(address)
      ? objKeys(meta)
      : (objDeepAction(meta, address, (parent, id) =>
          objKeys(parent[id] as Meta),
        ) ?? []);

  const readTimestamps = async (address: Address, _context: Context) =>
    objDeepAction(
      meta,
      address,
      (parent, id) => parent[id] ?? {},
    ) as Timestamps;

  const connector = await createMetaConnector(
    depth,
    {readTimestamp, writeTimestamp, readChildIds, readTimestamps},
    {getMeta: async () => jsonParse(jsonString(meta))},
  );

  return connector;
};
