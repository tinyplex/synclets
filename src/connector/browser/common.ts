import type {Data, Meta} from '@synclets/@types';
import {
  LocalStorageDataConnector,
  LocalStorageMetaConnector,
  SessionStorageDataConnector,
  SessionStorageMetaConnector,
} from '@synclets/@types/connector/browser';
import {jsonParse, jsonString} from '@synclets/utils';
import {createMemoryConnector} from '../../common/memory.ts';
export {resolve} from 'path';

export const createStorageConnector = <
  CreateLocal extends boolean,
  CreateMeta extends boolean,
  Depth extends number,
>(
  createLocal: CreateLocal,
  createMeta: CreateMeta,
  depth: Depth,
  storageName: string,
) => {
  const storage = globalThis[createLocal ? 'localStorage' : 'sessionStorage'];

  const onChange = async (tree: CreateMeta extends true ? Meta : Data) =>
    storage.setItem(storageName, jsonString(tree));

  const getInitialAfterConnect = () => {
    try {
      return jsonParse(storage.getItem(storageName) ?? '');
    } catch {}
  };

  const extraFunctions = {
    getStorageName: () => storageName,
  };

  return createMemoryConnector(
    createMeta,
    depth,
    undefined,
    onChange,
    getInitialAfterConnect,
    extraFunctions,
  ) as CreateLocal extends true
    ? CreateMeta extends true
      ? LocalStorageMetaConnector<Depth>
      : LocalStorageDataConnector<Depth>
    : CreateMeta extends true
      ? SessionStorageMetaConnector<Depth>
      : SessionStorageDataConnector<Depth>;
};
