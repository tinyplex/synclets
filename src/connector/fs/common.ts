import {createDataConnector, createMetaConnector} from '@synclets';
import type {
  Address,
  AnyParentAddress,
  Atom,
  AtomAddress,
  Data,
  Meta,
  Timestamp,
  TimestampAddress,
} from '@synclets/@types';
import {
  DirectoryDataConnector,
  DirectoryMetaConnector,
  FileDataConnector,
  FileMetaConnector,
} from '@synclets/@types/connector/fs';
import {isAtom, isTimestamp} from '@synclets/utils';
import {
  decodePaths,
  encodePaths,
  getDirectoryContents,
  readFileJson,
  removeFileAndAncestors,
  validateDirectory,
  validateFile,
  writeFileJson,
} from '../../common/fs.ts';
import {createMemoryConnector} from '../../common/memory.ts';

export const readLeaf = async <Leaf extends Atom | Timestamp>(
  directory: string,
  address: Address,
  isLeaf: (leaf: unknown) => leaf is Leaf | undefined,
): Promise<Leaf | undefined> => {
  const leaf = await readFileJson(directory, encodePaths(address));
  return isLeaf(leaf) ? leaf : undefined;
};

export const createFileConnector = <
  CreateMeta extends boolean,
  Depth extends number,
>(
  createMeta: CreateMeta,
  depth: Depth,
  file: string,
) => {
  let validatedFile: string;

  const connect = async () => {
    validatedFile = await validateFile(file);
  };

  const onChange = (tree: CreateMeta extends true ? Meta : Data) =>
    writeFileJson(validatedFile, [], tree, false);

  const getInitialAfterConnect = () =>
    readFileJson(validatedFile, []) as Promise<
      (CreateMeta extends true ? Meta : Data) | undefined
    >;

  const extraFunctions = {
    getFile: () => file,
  };

  return createMemoryConnector(
    createMeta,
    depth,
    connect,
    onChange,
    getInitialAfterConnect,
    extraFunctions,
  ) as CreateMeta extends true
    ? FileMetaConnector<Depth>
    : FileDataConnector<Depth>;
};

export const createDirectoryConnector = <
  CreateMeta extends boolean,
  Depth extends number,
>(
  createMeta: CreateMeta,
  depth: Depth,
  directory: string,
) => {
  let validatedDirectory: string;

  const connect = async () => {
    validatedDirectory = await validateDirectory(directory);
  };

  const readAtom = (address: AtomAddress<Depth>): Promise<Atom | undefined> =>
    readLeaf(validatedDirectory, address, isAtom);

  const readTimestamp = (
    address: TimestampAddress<Depth>,
  ): Promise<Timestamp | undefined> =>
    readLeaf(validatedDirectory, address, isTimestamp);

  const writeLeaf = (
    address: AtomAddress<Depth> | TimestampAddress<Depth>,
    leaf: Atom | Timestamp,
  ) => writeFileJson(validatedDirectory, encodePaths(address), leaf);

  const removeAtom = (address: AtomAddress<Depth>) =>
    removeFileAndAncestors(validatedDirectory, encodePaths(address));

  const readChildIds = async (address: AnyParentAddress<Depth>) =>
    decodePaths(
      await getDirectoryContents(validatedDirectory, encodePaths(address)),
    );

  const extraFunctions = {
    getDirectory: () => directory,
  };

  const connector = createMeta
    ? createMetaConnector(
        depth,
        {
          connect,
          readTimestamp,
          writeTimestamp: writeLeaf,
          readChildIds,
        },
        {},
        extraFunctions,
      )
    : createDataConnector(
        depth,
        {
          connect,
          readAtom,
          writeAtom: writeLeaf,
          removeAtom,
          readChildIds,
        },
        {},
        extraFunctions,
      );

  return connector as CreateMeta extends true
    ? DirectoryMetaConnector<Depth>
    : DirectoryDataConnector<Depth>;
};
