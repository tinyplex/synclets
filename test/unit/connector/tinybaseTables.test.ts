import {createSynclet} from 'synclets';
import {createMemoryMetaConnector} from 'synclets/connector/memory';
import {createTinyBaseTablesDataConnector} from 'synclets/connector/tinybase';
import {createMemoryTransport} from 'synclets/transport/memory';
import {createStore} from 'tinybase/store';
import {expect, test} from 'vitest';
import {describeSyncletTests} from '../common.ts';

describeSyncletTests(
  'memory',
  async () => {},
  async () => {},
  () => createTinyBaseTablesDataConnector(createStore()),
  (depth: number) => createMemoryMetaConnector(depth),
  (uniqueId: string) => createMemoryTransport({poolId: uniqueId}),
  1,
  [3],
);

test('getStore, tables', async () => {
  const store = createStore();
  const dataConnector = createTinyBaseTablesDataConnector(store);
  const synclet = await createSynclet({dataConnector});
  expect(dataConnector.getStore()).toBe(store);
  expect(synclet.getDataConnector().getStore()).toBe(store);
});
