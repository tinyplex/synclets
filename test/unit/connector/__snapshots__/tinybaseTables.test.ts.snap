// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`over memory > 3-depth > 2-way > connected > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > connected > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > connected, deletion > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > connected, deletion > equivalent 2`] = `{}`;

exports[`over memory > 3-depth > 2-way > connected, far atom > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
  "b": {
    "ba": {
      "baa": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > connected, initial > equivalent 1`] = `{}`;

exports[`over memory > 3-depth > 2-way > connected, near atom > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
      "aab": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > disconnected, conflicting values > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > disconnected, conflicting values 2 > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
      "aab": "C",
    },
  },
  "b": {
    "ba": {
      "baa": "D",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > disconnected, far atom > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
  "b": {
    "ba": {
      "baa": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > disconnected, near atom > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
      "aab": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > set 1, set 2, start 2, start 1 > differing 1`] = `
[
  {
    "a": {
      "aa": {
        "aaa": "A",
      },
    },
  },
  {},
]
`;

exports[`over memory > 3-depth > 2-way > set 1, set 2, start 2, start 1 > differing 2`] = `
[
  {
    "a": {
      "aa": {
        "aaa": "A",
      },
    },
  },
  {
    "a": {
      "aa": {
        "aaa": "B",
      },
    },
  },
]
`;

exports[`over memory > 3-depth > 2-way > set 1, set 2, start 2, start 1 > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > start 1, set 1, start 2 > differing 1`] = `
[
  {
    "a": {
      "aa": {
        "aaa": "A",
      },
    },
  },
  {},
]
`;

exports[`over memory > 3-depth > 2-way > start 1, set 1, start 2 > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > start 2, set 1, start 1 > differing 1`] = `
[
  {
    "a": {
      "aa": {
        "aaa": "A",
      },
    },
  },
  {},
]
`;

exports[`over memory > 3-depth > 2-way > start 2, set 1, start 1 > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > stop 1, set 1, start 1 > differing 1`] = `
[
  {
    "a": {
      "aa": {
        "aaa": "B",
      },
    },
  },
  {
    "a": {
      "aa": {
        "aaa": "A",
      },
    },
  },
]
`;

exports[`over memory > 3-depth > 2-way > stop 1, set 1, start 1 > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > stop 1, set 1, start 1 > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > stop 1, set 2, start 1 > differing 1`] = `
[
  {
    "a": {
      "aa": {
        "aaa": "A",
      },
    },
  },
  {
    "a": {
      "aa": {
        "aaa": "B",
      },
    },
  },
]
`;

exports[`over memory > 3-depth > 2-way > stop 1, set 2, start 1 > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A",
    },
  },
}
`;

exports[`over memory > 3-depth > 2-way > stop 1, set 2, start 1 > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "B",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > chain > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A1",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > chain > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "A2",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > chain > equivalent 3`] = `
{
  "a": {
    "aa": {
      "aaa": "A3",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > pool > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A1",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > pool > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "A2",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > pool > equivalent 3`] = `
{
  "a": {
    "aa": {
      "aaa": "A3",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > ring > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A1",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > ring > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "A2",
    },
  },
}
`;

exports[`over memory > 3-depth > 3-way > ring > equivalent 3`] = `
{
  "a": {
    "aa": {
      "aaa": "A3",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A1",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "A2",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 3`] = `
{
  "a": {
    "aa": {
      "aaa": "A3",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 4`] = `
{
  "a": {
    "aa": {
      "aaa": "A4",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 5`] = `
{
  "a": {
    "aa": {
      "aaa": "A5",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 6`] = `
{
  "a": {
    "aa": {
      "aaa": "A6",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 7`] = `
{
  "a": {
    "aa": {
      "aaa": "A7",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 8`] = `
{
  "a": {
    "aa": {
      "aaa": "A8",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 9`] = `
{
  "a": {
    "aa": {
      "aaa": "A9",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > chain > equivalent 10`] = `
{
  "a": {
    "aa": {
      "aaa": "A10",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A1",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "A2",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 3`] = `
{
  "a": {
    "aa": {
      "aaa": "A3",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 4`] = `
{
  "a": {
    "aa": {
      "aaa": "A4",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 5`] = `
{
  "a": {
    "aa": {
      "aaa": "A5",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 6`] = `
{
  "a": {
    "aa": {
      "aaa": "A6",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 7`] = `
{
  "a": {
    "aa": {
      "aaa": "A7",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 8`] = `
{
  "a": {
    "aa": {
      "aaa": "A8",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 9`] = `
{
  "a": {
    "aa": {
      "aaa": "A9",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > pool > equivalent 10`] = `
{
  "a": {
    "aa": {
      "aaa": "A10",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 1`] = `
{
  "a": {
    "aa": {
      "aaa": "A1",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 2`] = `
{
  "a": {
    "aa": {
      "aaa": "A2",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 3`] = `
{
  "a": {
    "aa": {
      "aaa": "A3",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 4`] = `
{
  "a": {
    "aa": {
      "aaa": "A4",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 5`] = `
{
  "a": {
    "aa": {
      "aaa": "A5",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 6`] = `
{
  "a": {
    "aa": {
      "aaa": "A6",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 7`] = `
{
  "a": {
    "aa": {
      "aaa": "A7",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 8`] = `
{
  "a": {
    "aa": {
      "aaa": "A8",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 9`] = `
{
  "a": {
    "aa": {
      "aaa": "A9",
    },
  },
}
`;

exports[`over memory > 3-depth > 10-way > ring > equivalent 10`] = `
{
  "a": {
    "aa": {
      "aaa": "A10",
    },
  },
}
`;
